### PR TITLE
Add metadata-finalized hooks in ComponentTrainer

### PR DESCRIPTION
**Pull request by mallamanis**

### CODE REVIEW

**Comment by mallamanis** [comment]
@ASvyatkovskiy see 520835c37cf58efdfd6e5406323804645b380403 of this PR.

**Comment by ASvyatkovskiy** [comment]
@mallamanis can training start hook be of type:
```python
Callable[[BaseComponent, torch.optim.Optimizer], None]
```
or ```Optional[torch.optim.Optimizer]```.

this would allow broadcasting both model parameters and optimizer state from root to all other processes during distributed training. E.g., with horovod:
```python
    def distributed_training_hook(model: BaseComponent, optimizer: torch.optim.Optimizer) -> None:
        hvd.broadcast_parameters(model.state_dict(), root_rank=0)
        hvd.broadcast_optimizer_state(optimizer, root_rank=0)
```

File: python/dpu_utils/ptutils/componenttrainer.py [hunk 1]
```diff
@@ -150,6 +167,9 @@ def minibatch_iterator(data_iterator: Iterator[TensorizedData], return_partial_m
                     mb_loss.backward()
 
                     optimizer.step()
+                    if scheduler is not None:
+                        scheduler.step(epoch)
```
Comment (mmjb) [inline python/dpu_utils/ptutils/componenttrainer.py]: Would it be helpful to change the interface to not only provide the `epoch`, but also the step index (would require an enumerate in the loop on L162)? This would make sense for things like warmup schedules and similar things, which are usually not epoch-based.

**Comment by mallamanis** [comment]
I've updated the code based on @mmjb and @ASvyatkovskiy comments:

* Training-Start hooks now have the interface `Callable[[BaseComponent, torch.optim.Optimizer], None]`
* The scheduler interface is defined in `AbstractScheduler` and can wrap possible schedulers. This is a necessary compromise, since some schedulers are to be invoked every time the _epoch_ changes, whereas others (like the huggingface ones) are meant to be called per training step.
